thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/alert/content/EntityGroupByContentFormatter.java

```diff
@@ -0,0 +1,140 @@
+
+  private static final String EMAIL_TEMPLATE = "emailTemplate";
+  private static final String DEFAULT_EMAIL_TEMPLATE = "entity-groupby-anomaly-report.ftl";
+
+  static final String PROP_ENTITY_NAME = "entityName";
+  static final String PROP_ANOMALY_SCORE = "groupScore";
+  static final String PROP_GROUP_KEY = "groupKey";
+
+  private DetectionConfigManager configDAO = null;
+  private Map<String, Double> entityAnomalyToScoreMap = new HashMap<>();
+  private Multimap<String, AnomalyReportEntity> entityAnomalyReports = ArrayListMultimap.create();
+  private Map<String, String> entityToAnomalyIdsMap = new HashMap<>();
+
+  public EntityGroupByContentFormatter() {}
+
+  @Override
+  public void init(Properties properties, EmailContentFormatterConfiguration configuration) {
+    super.init(properties, configuration);
+    this.emailTemplate = properties.getProperty(EMAIL_TEMPLATE, DEFAULT_EMAIL_TEMPLATE);
+    this.configDAO = DAORegistry.getInstance().getDetectionConfigManager();
+  }
+
+  @Override
+  protected void updateTemplateDataByAnomalyResults(Map<String, Object> templateData,
+      Collection<AnomalyResult> anomalies, EmailContentFormatterContext context) {
+    DetectionConfigDTO config = null;
+    Preconditions.checkArgument(anomalies != null && !anomalies.isEmpty(), "Report has empty anomalies");
+    for (AnomalyResult anomalyResult : anomalies) {
+      if (!(anomalyResult instanceof MergedAnomalyResultDTO)) {
+        LOG.warn("Anomaly result {} isn't an instance of MergedAnomalyResultDTO. Skip from alert.", anomalyResult);
+        continue;
+      }
+
+      MergedAnomalyResultDTO anomaly = (MergedAnomalyResultDTO) anomalyResult;
+      if (config == null) {
+        config = this.configDAO.findById(anomaly.getDetectionConfigId());
+        Preconditions.checkNotNull(config, String.format("Cannot find detection config %d", anomaly.getDetectionConfigId()));
+      }
+
+      updateEntityToAnomalyDetailsMap(anomaly, config);
+    }
+
+    List<Map.Entry<String, Double>> anomaliesSortedByScores = new ArrayList<>(entityAnomalyToScoreMap.entrySet());
+    anomaliesSortedByScores.sort(Map.Entry.comparingByValue());
+    List<String> entityList = anomaliesSortedByScores.stream().map(Map.Entry::getKey).collect(Collectors.toList());
+    entityList.sort(Collections.reverseOrder());
+
+    templateData.put("emailHeading", config.getName());
+    templateData.put("entityToAnomalyIdsMap", entityToAnomalyIdsMap);
+    templateData.put("entitySortedByScoreList", entityList);
+    templateData.put("entityToAnomalyDetailsMap", entityAnomalyReports.asMap());
+  }
+
+  /**
+   * Recursively find the anomalies having a groupKey and display them in the email
+   */
+  private void updateEntityToAnomalyDetailsMap(MergedAnomalyResultDTO anomaly, DetectionConfigDTO detectionConfig) {
+    if (anomaly.getProperties() != null && anomaly.getProperties().containsKey(PROP_GROUP_KEY) && anomaly.getChildIds() != null) {
```

Comment:
Should it be `getChildIds().isEmpty()`. ?

Reply:
We need the null check anyways. I refactored this portion a bit and moved the check below.